[AArch64] Optimization of repeated constant loads (#51483)

ParkHanbum opened the pull request:
This change looks for cases where load same constant in the upper and lower 32-bit sizes of a 64-bit  space. If we found, it loads only the lower 32-bit constant and replaces it with an instruction that stores it simultaneously in the upper and lower.

For example:
  renamable $x8 = MOVZXi 49370, 0
  renamable $x8 = MOVKXi $x8, 320, 16
  renamable $x8 = MOVKXi $x8, 49370, 32
  renamable $x8 = MOVKXi $x8, 320, 48
  STRXui killed renamable $x8, killed renamable $x0, 0
becomes
  $w8 = MOVZWi 49370, 0
  $w8 = MOVKWi $w8, 320, 16
  STPWi killed renamable $w8, killed renamable $w8, killed renamable $x0, 0

davemgreen commented:
It looks like gcc will transform repeated constants to `or x, x, lsl 32` too:
https://gcc.godbolt.org/z/h6MqTja1s

Is it worth doing that first, then this could be simplified a little to just having to look at `str(or(x, x, lsl 32))`.

Can you add some testing too? Thanks

ParkHanbum commented:
@davemgreen sure, consider it done!

davemgreen commented:
Thanks! I think I would expect it to be an additional case in AArch64_IMM::expandMOVImm(..), that can check the two halves are the same.

ParkHanbum commented:
thanks for tip!

ParkHanbum commented:
@davemgreen Do we support the same ORR format as GCC? 
I get an error when I give ORR three registers as Operand, I wonder if this is because it doesn't support Instruction Format.
https://gcc.godbolt.org/z/131YsEY6W

```
def : InstAlias<"mov $dst, $imm", (ORRWri GPR32sp:$dst, WZR,
                                          logical_imm32:$imm), 0>;
def : InstAlias<"mov $dst, $imm", (ORRXri GPR64sp:$dst, XZR,
                                          logical_imm64:$imm), 0>;
```

if I need to add new instruction format, let me know!

ParkHanbum commented:
@davemgreen Do we support the same ORR format as GCC? 
I get an error when I give ORR three registers as Operand, I wonder if this is because it doesn't support Instruction Format.
https://gcc.godbolt.org/z/131YsEY6W

could you little advise which instruction can I use please?

davemgreen commented:
Hi - I believe you will need a ORRXrs. The ORRXri will take an immediate for the second operand.

It's like one of these: https://gcc.godbolt.org/z/8aWhfWv4T (but in that case might need to check that the top bits are known to be 0 when we convert it to a stp).

ParkHanbum commented:
@davemgreen :) thanks!

efriedma-quic commented:
> Is it worth doing that first, then this could be simplified a little to just having to look at str(or(x, x, lsl 32)).

This handles some cases, but not all.  Consider:

```
void f(long long *A) {
   *A = 0xC2000000C2000000;
}

long long f2() {
    return 0xC2000000C2000000;
}
```

You want mov+stp for the store... but you don't want to generate orr for the return.  (Even for the cases where we generate three instructions, we're going to avoid "orr(x, x, lsl 32)" where we can because it's more expensive than movk.)

Maybe we should run this optimization before pseudo-expansion?

ParkHanbum commented:
@efriedma-quic it is good for me.
what I trying to now is if we found ORR from sequence of backward then do Bit-OR when match "movn" finally.

like this. 

```
    // ValueOrder == 1 mean this is MOVZXi. 
    if (ValueOrder == 1 && DupBitSize) {
      Accumulated |= Accumulated << DupBitSize;
      DupBitSize = 0;
    } 
 ```

but this is base on assuming that "ORR(x, x, lsl 32)" was generated in pseudo-expansion. 
so, it is need verification by you experts.

davemgreen commented:
> > Is it worth doing that first, then this could be simplified a little to just having to look at str(or(x, x, lsl 32)).
> 
> This handles some cases, but not all. Consider:
> 
> ```
> void f(long long *A) {
>    *A = 0xC2000000C2000000;
> }
> 
> long long f2() {
>     return 0xC2000000C2000000;
> }
> ```
> 
> You want mov+stp for the store... but you don't want to generate orr for the return. (Even for the cases where we generate three instructions, we're going to avoid "orr(x, x, lsl 32)" where we can because it's more expensive than movk.)
> 
> Maybe we should run this optimization before pseudo-expansion?

`orr(x, x, lsl n)` is usually as cheap as a movk. It's really just a `lsl #n` with an `or` on the end, and unlike add/sub it can be done in a single cycle. Unfortunately that might not apply for every cpu, but should be for everything modern enough.

You might be right that we might need both methods in the end though. I was hoping we could add the `orr` combine because it is useful on it's own.

efriedma-quic commented:
> orr(x, x, lsl n) is usually as cheap as a movk.

On which processors?  It looks like on Cortex-X2 and newer, "orr(x, x, lsl n)" has the same cost as movk, but looking through various sources, I can't find any other chip that does.  Granted, it's probably still better than generating two MOVK.

efriedma-quic commented:
> but this is base on assuming that "ORR(x, x, lsl 32)" was generated in pseudo-expansion.

I think it's fine to assume this as a starting-point; we can try to refine it in a followup.

ParkHanbum commented:
is there anyone who can review my code?

ParkHanbum commented:
can I ask code review??

davemgreen commented:
Hello. It would be best if there was one patch that added support for or(x, x, lsl 32), so we can make sure that is tested and committed and doesn't cause any problems. The load/store optimizer change should then be an addition on top.

Whenever we change these mir optimization passes there seems to be a high chance of subtle things going wrong.

ParkHanbum commented:
@davemgreen got it, I'll split this. 
@efriedma-quic I didn't know it. what can I do for request review?

efriedma-quic commented:
Just leave a comment on the pull request noting that you've updated the patch.

(Actually, I've seen GitHub generates notification for new commits in some cases, but in this case it didn't generate a notification; not sure what the difference is.)

ParkHanbum committed:
[AArch64] Optimize `MOV` to `ORR` when load symmetric constants (#51483)

This change looks for cases of symmetric constant loading.
`symmetric constant load` is when the upper 32 bits and lower 32 bits
of a 64-bit register load the same value.

When it finds this, it replaces it with an instruction that loads only
the lower 32 bits of the constant and stores it in the upper and lower
bits simultaneously.

For example:
  renamable $x8 = MOVZXi 49370, 0
  renamable $x8 = MOVKXi $x8, 320, 16
  renamable $x8 = MOVKXi $x8, 49370, 32
  renamable $x8 = MOVKXi $x8, 320, 48
becomes
  renamable $x8 = MOVZXi 49370, 0
  renamable $x8 = MOVKXi $x8, 320, 16
  renamable $x8 = ORRXrs $x8, $x8, 32

ParkHanbum committed:
[AArch64] Add PreTest for optimizing `MOV` to `ORR`

davemgreen commented:
Can you make the first half of this into a separate pr? Thanks

ParkHanbum commented:
@davemgreen sorry I misunderstood. The test case I created consists of only one MOV Instruction. I'm not sure how to create a failure case in this case, could you please advise?

davemgreen commented on llvm/lib/Target/AArch64/AArch64ExpandImm.cpp:
```diff
@@ -509,6 +509,16 @@ static inline void expandMOVImmSimple(uint64_t Imm, unsigned BitSize,
+    Insn.push_back(
+        {Opc, Imm16, AArch64_AM::getShifterImm(AArch64_AM::LSL, Shift)});
+  if (Imm != 0 && (Imm >> 32) == (Imm & UINT_MAX)) {
+    Insn.push_back({BitSize == 32 ? AArch64::ORRWrs : AArch64::ORRXrs, 0, 32});
```
If BitSize == 32 is unreachable then it might be worth only adding AArch64::ORRXrs.

davemgreen reviewed (commented):
Thanks, this is looking decent to me. I think it is probably worth adding .ll file tests to check that everything works together properly. Maybe into a test like movw-const.ll.

These are some of the ones I tried. It would be good if 0x0000555500005555 still produced a mov+movk, and the last one is failing probably because of the f's. I'm not sure if there are other patterns that would be worth testing too?
```
define i64 @test0x1234567812345678() {
; CHECK-LABEL: test0x1234567812345678:
; CHECK:       // %bb.0:
; CHECK-NEXT:    mov x0, #22136 // =0x5678
; CHECK-NEXT:    movk x0, #4660, lsl #16
; CHECK-NEXT:    orr x0, x0, x0, lsl #32
; CHECK-NEXT:    ret
  ret i64 u0x1234567812345678
}

define i64 @test0xff3456ffff3456ff() {
; CHECK-LABEL: test0xff3456ffff3456ff:
; CHECK:       // %bb.0:
; CHECK-NEXT:    mov x0, #22271 // =0x56ff
; CHECK-NEXT:    movk x0, #65332, lsl #16
; CHECK-NEXT:    orr x0, x0, x0, lsl #32
; CHECK-NEXT:    ret
  ret i64 u0xff3456ffff3456ff
}

define i64 @test0x00345600345600() {
; CHECK-LABEL: test0x00345600345600:
; CHECK:       // %bb.0:
; CHECK-NEXT:    mov x0, #22016 // =0x5600
; CHECK-NEXT:    movk x0, #52, lsl #16
; CHECK-NEXT:    movk x0, #13398, lsl #32
; CHECK-NEXT:    ret
  ret i64 u0x00345600345600
}

define i64 @test0x5555555555555555() {
; CHECK-LABEL: test0x5555555555555555:
; CHECK:       // %bb.0:
; CHECK-NEXT:    mov x0, #6148914691236517205 // =0x5555555555555555
; CHECK-NEXT:    ret
  ret i64 u0x5555555555555555
}

define i64 @test0x5055555550555555() {
; CHECK-LABEL: test0x5055555550555555:
; CHECK:       // %bb.0:
; CHECK-NEXT:    mov x0, #6148914691236517205 // =0x5555555555555555
; CHECK-NEXT:    and x0, x0, #0xf0fffffff0ffffff
; CHECK-NEXT:    ret
  ret i64 u0x5055555550555555
}

define i64 @test0x0000555555555555() {
; CHECK-LABEL: test0x0000555555555555:
; CHECK:       // %bb.0:
; CHECK-NEXT:    mov x0, #6148914691236517205 // =0x5555555555555555
; CHECK-NEXT:    movk x0, #0, lsl #48
; CHECK-NEXT:    ret
  ret i64 u0x0000555555555555
}

define i64 @test0x0000555500005555() {
; CHECK-LABEL: test0x0000555500005555:
; CHECK:       // %bb.0:
; CHECK-NEXT:    mov x0, #21845 // =0x5555
; CHECK-NEXT:    orr x0, x0, x0, lsl #32
; CHECK-NEXT:    ret
  ret i64 u0x0000555500005555
}

define i64 @testu0xffff5555ffff5555() {
; CHECK-LABEL: testu0xffff5555ffff5555:
; CHECK:       // %bb.0:
; CHECK-NEXT:    mov x0, #-43691 // =0xffffffffffff5555
; CHECK-NEXT:    orr x0, x0, x0, lsl #32
; CHECK-NEXT:    ret
  ret i64 u0xffff5555ffff5555
}
```

I looked through the other uses of expandMOVImm and they all check Insn.size() == 1 or are only checking for the number of instructions.

davemgreen reviewed (commented):
Thanks for splitting out the tests, they look good apart for the one with ffs where isNeg=true.

ParkHanbum commented on llvm/test/CodeGen/AArch64/movimm-expand-ldst.ll:
```diff
@@ -0,0 +1,76 @@
+; CHECK-LABEL: testu0xffff5555ffff5555:
+; CHECK:       // %bb.0:
+; CHECK-NEXT:    mov x0, #-43691 // =0xffffffffffff5555
+; CHECK-NEXT:    orr x0, x0, x0, lsl #32
```
Thank you very much for showing us how to write tests. I added two tests

davemgreen commented on llvm/lib/Target/AArch64/AArch64ExpandImm.cpp:
```diff
@@ -518,6 +518,14 @@ static inline void expandMOVImmSimple(uint64_t Imm, unsigned BitSize,
+
+  // Now, we get 16-bit divided Imm. If high and low bits are same in
+  // 32-bit, there is an opportunity to reduce instruction.
+  if (Insn.size() > 2 && (Imm >> 32) == (Imm & UINT_MAX)) {
```
I would use 0xffffffff instead of UINT_MAX. They are likely always the same value, but the ff's is more explicit about the size.

ParkHanbum replied:
got it!

davemgreen reviewed (commented):
Thanks - I gave this a test and it did OK. It came up a decent amount, and I couldn't see any problems. Other than changing the UINT_MAX this LGTM.

davemgreen commented:
Thanks, LGTM

davemgreen commented:
Can you update the commit message? Then let me know if I should hit submit

davemgreen commented:
Pushed in ce1a0d8ad380d12dc7ea001cfab3749bb23d445d.
